### ISSUE TITLE
SImplify authorization definitions?

### ISSUE THREAD

**Issue by kevinrobinson** [studentinsights/studentinsights#1080]
This is a thought from https://github.com/studentinsights/studentinsights/issues/1078.

re: authorization... It might be helpful to clarify some of the authorization in code (trying to remove some of the nuance there like "educators who have permission to view a student's profile will be able to see the student's grade in all current year sections, regardless of whether or not they have permission to view the sections".  It looks like this was modeled after the homerooms code, and I think with increasing complexity in these rules maybe a good opportunity to simplify.

One simplification might be: authorization is done on a `(educator, student)` basis, full stop ([here](https://github.com/studentinsights/studentinsights/blob/master/app/models/educator.rb#L63)).  And other things like "can they see a homeroom?" or "can they see a section?" are not authorization or security features, but convenience.  In other words, what if https://github.com/studentinsights/studentinsights/blob/8e7da6300f40109c1e953606308422af85ec29d1/app/controllers/sections_controller.rb#L27 showed anyone who asked information overall about the session (including the total number of students), but then for each student in that session, asked if the educator was authorized and only showed those students' data.

This same approach would apply to Homerooms as well, [here](https://github.com/studentinsights/studentinsights/blob/master/app/controllers/homerooms_controller.rb#L58).  Basically whenever we're looking up student data in controller code, we run it through the authorization check based on the current educator.  We could do a sneaky Rails thing to make that happen automagically, but I'd be in favor of doing it explicitly.

Concretely, anyone could look at any section/homeroom, but they'd only see students they're authorized to see.  Asking for "their sections" or "their homerooms" would be a convenience question and not an authorization and data security question.  The main benefit is a simpler model for authorization, where we have only one method that controls access to student data instead of three that are slightly duplicative and likely to drift out of sync.  The cost might be having to index or precompute some of that data if it's two slow to compute the access for each student at read time, but that's just a guess.

**Comment by jhilde** [comment]
I kinda see where you're going but I'm not sure about everyone being able to see all homerooms and sections with only students they have permission to see. When you view a section or homeroom, you're likely not thinking about a particular student...you're thinking about a grouping of students. Showing some but not all section or homeroom students seems incomplete.

Also, the majority of users gain permission not through a relationship with students but a relationship to a section or homeroom. 

That being said, I have contemplated moving the logic of which students (and perhaps sections and homerooms) a user can see to be precomputed. That would centralize the more complicated logic of the ways one can gain access to a student. And open up the possibility of allowing an admin to grant ad hoc access to an individual student. 

The other interesting view I think would be a roster of all students a user has access to. So essentially an educator dashboard that provides access to drill down via their sections, homerooms, and students. I think this is especially relevant for high school teachers who are likely teaching many sections. Currently they log in and see the first section within the database and a drop down. I think a dashboard makes sense for them.

**Comment by kevinrobinson** [comment]
@jhilde Thanks, it's great to talk this out!  :)  The main thing I think is important is having one place that defines the authorization rules for data access.  It's already quite complicated, and bugs there are among the most costly we can make since they lead to unauthorized access to student data.

Also, re: gain permission - for me the data security and authorization is what's important, and not the user flows or pages in the product.  Right now, this is handled in different ways in the product, and so it's easy for the authorization rules are subtly different depending on what URLs are requested.  Concretely, my main suggestion for this issue is that we aim to unify it so that these all use the same rules for determining authorization:

- [student profile](https://github.com/studentinsights/studentinsights/blob/master/app/controllers/students_controller.rb#L16)
- [homeroom](https://github.com/studentinsights/studentinsights/blob/master/app/controllers/homerooms_controller.rb#L36)
- [section](https://github.com/studentinsights/studentinsights/blob/master/app/controllers/sections_controller.rb#L13)

Ideally we'd have those authorization rules be totally isolated from changes to the product experience and UX, and product designs wouldn't have to consider this.  We could have https://github.com/studentinsights/studentinsights/blob/master/app/models/educator.rb#L61 be that place, or some other place, but I think having it in one place will help us reduce the risk of authorization bugs.

**Comment by kevinrobinson** [comment]
See also: https://github.com/studentinsights/studentinsights/pull/1127/files#diff-d7885bcb31c3967f15650a124edd67b4R15

**Comment by kevinrobinson** [comment]
Another spot: https://github.com/studentinsights/studentinsights/blob/master/app/jobs/precompute_student_hashes_job.rb#L27

**Comment by kevinrobinson** [comment]
From @alexsoble on https://github.com/studentinsights/studentinsights/pull/1127#issuecomment-333886223:

> @kevinrobinson, can you help me understand your ideas about authorization? The reason the precomputing code I wrote asked about educator authorization is so that it could surface only students the educator is authorized to see in the searchbar.

> Are you saying we don't want the searchbar code to be aware of authorization at all? So that (for example) any educator would get the entire list of students in the district when they type in the searchbar, but get a "You are not authorized" warning from the server side if they try to visit the profile of a student they're not authorized for?

To answer your questions: yes, I'm suggesting that the searchbar code shouldn't be aware of authorization, in the sense that it shouldn't have any logic or code related to authorization but no, I'm not suggesting that we send down all students to all educators.  Specifically in https://github.com/studentinsights/studentinsights/pull/1127/files#r142385177 I'm suggesting we avoid adding code to features that checks educator permissions, and funnel all those decisions through a central place.  So that for the searchbar, conceptually we'd call `Student.all.map {|s| current_educator.is_authorized_for_student(s) }` and that would tell us what students to show.  We wouldn't write code in the searchbar feature that figures out which students to show, or checks permissions on educators.  That's not part of the searchbar feature, it's part of the authorization rules.

More broadly, the question is: when writing Ruby code, how do we find out who has access to a student's data?  Right now this is quite complicated, and defined in a few different places that can easily get out of sync (see the links above).  I'm suggesting that we answer that question in one place, `Educator#is_authorized_for_student` or in some other method or class.  If we need helper methods like "what students can this educator see?" that those helper methods would ultimately call into one set of rules (eg, `Educator#is_authorized_for_student`).  The variations of "what sections can this educator see?" and "what homerooms can this educator see?" have their own complexity, and I'm wondering if we can centralize all this.  One way to get a sense of this complexity is to run the `simulate` code in [this PR](https://github.com/studentinsights/studentinsights/pull/1129) and start asking why so many educators don't have access to any students, and if that is correct or not.  It's complicated, and depending on what way you asked right now, you could get different answers.

The idea that motivates that is that changing authorization code is risky, so we should strive to move it into one place that we know is risky to change, be careful changing it, and then not have to think about it when shipping most product features.  So we separate the concern of "can this educator see this student's data?" from "I want a searchbar that lets educators jump to a student's profile page."

Overall, these are great things to talk out and improve!  My impression is that there's nothing blocking this from happening in the links above.  For the searchbar, we ran into a complication because the nightly job took too long.  If that's what's blocking it, we could profile why it takes so long to compute the authorization and optimize that.  I'm not tied to any of the particulars here other than the general principle of "put authorization rules in one place."  I'd be excited about any ways we can take steps towards that. :)

**Comment by alexsoble** [comment]
Ooooooh I understand now. Thanks for the longer explanation @kevinrobinson.
Just going to try to rewrite it in my own words to make sure I
understand... Lots of parts of the app need to know about authorization.
But they shouldn't each write their own auth checks. Because that gets
confusing and creates surface for more mistakes. They should all call one
central auth check instead.

Thanks for being patient with me in grasping this concept! :)

**Comment by kevinrobinson** [comment]
Of course!  This is tricky and I don't know the best way to approach it.  @alexsoble @jhilde If you all have other thoughts on this that'd be awesome too.

**Comment by kevinrobinson** [comment]
Here's a notional sketch of what this might look like: https://gist.github.com/kevinrobinson/02c0f53f243c2812f774f0a3eae61dc8.  If you all have thoughts on that, other ideas, or feedback that this isn't as important and I should focus on X instead, that's all great!

Also, gist comments don't come out as emails, so comments here on the issue are probably better.

**Comment by kevinrobinson** [comment]
Another related place: educators_controller#filtered_names (which is being passed a student_id so gives an indirect signal about services provided to other students at the school.  this is probably fine, but should tighten this as well).

**Comment by alexsoble** [comment]
Hey @kevinrobinson, @jhilde — how do we feel about adding the **critical** label to this issue and setting it as one of our highest priorities in the "Strengthen Code that's Vulnerable to Breaking" category? 

Of all the issues we're thinking through and working on, this touches the most vital topic: correctly assigning and interpreting educator permissions. And the longer we wait to ship this, the more code we'll have to refactor. 

Shipping this will also make it much easier for contributors to add new dashboard pages because they won't have to worry about writing authorization code.

**Comment by kevinrobinson** [comment]
Sounds good to me, I'm also happy to handoff if other folks can tackle this in the more immediate term.

Previously I looked at Pundit, CanCan and CanCanCan to see if they were good fits.  As I remember, the answer was mostly no, because there was too much Rails-style magic, it was too tied to controller code in particular (which doesn't work for precomputing) or that it didn't have particularly good support for collections-related authorization.  I could share more about this, or other folks could look more at these too.

The branch I was working on is here: https://github.com/studentinsights/studentinsights/compare/proposal/authorization (oops, it's a few commits behind though).  This was looking at:
- Can we use `default_scope` on a model to restrict direct model-layer access and enforce that all calls go through an authorization-aware class?  The answer is "this is sort of helpful for development and tests, but leads to behavior that is too counter-intuitive to be shipped in production."  See [here](https://github.com/studentinsights/studentinsights/compare/proposal/authorization#diff-29ff81d6af0261f9e9a494a7fbd5fa08R15) and [here](https://github.com/studentinsights/studentinsights/compare/proposal/authorization#diff-02eba7263bd69fb5d40aea67aa6bcfdfR98) if you're curious, but I think we should not do this.
- Can we return lazily-evaluated `ActiveRecord::Relation` objects that have Ruby authorization code applied to them?  Not really, once you run `select` or Ruby-level filtering on a `Relation`, it forces eager evaluation.  That means that we can't pass around "the list of authorized students" without eagerly evaluating that, which is probably no bueno for performance.  See [here](https://github.com/studentinsights/studentinsights/compare/proposal/authorization#diff-af9715add29bf15dd8a7ee106f0c1c9aR33) for where I was looking at doing this and notes in [these comments](https://github.com/studentinsights/studentinsights/compare/proposal/authorization#diff-af9715add29bf15dd8a7ee106f0c1c9aR17).
- What's the right API here?  oops, this is missing the latest commits.  But where I last landed was something like:

```
students = guard { Student.find(params[:id]) }
students = guard { Homeroom.find(params[:id]).students }
sections = guard { Student.find(params[:id]).sections }
```
where guard is:

```
class Guard
  def initialize(educator)
    @educator = educator
  end
  def guard(&block)
    ...
  end
end
```

and in the `#guard` method it runs the block, then for each model, Relation or array that's returned, it applies authorization code to filter out the response.  The benefit here is that we have a single-method privacy-aware API for developers, we don't have to re-implement the ActiveRecord API, and the implementation of the method just branches on class type.  We can add a check in the controller layer to see that `guard` is called by each endpoint.

As a first cut, I was aiming to look at access to the Student model (via `#find` or via other associations) and see what that looks like.  Then reflect and discuss if we want to go forward with that for other models too.

I'm open to any other suggestions, or if other folks want to grab this and move it along faster that's great too!

**Comment by jhilde** [comment]
Looks like an interesting approach. So that guard function would essentially define all auth code for the app and act like a governor for student data. 

So for instance, if I wanted the list of students an educator could access, would I still duplicate the logic (i.e. homeroom, sections, school wide etc) or would I just allow the guard function to filter the entire list of students?

Would the logic live in 2 places? Guard and the controller? Guard is a check to make sure nothing accidentally gets out and the controller logic improves performance by not grabbing the entire list of students?

Back in the old days, we would do this with database views and stored procs, effectively creating a custom ORM for developers. But yes, agree with @kevinrobinson, let's not recreate ActiveRecord.

**Comment by kevinrobinson** [comment]
Yeah the overall idea is to just use the guard filter (and not duplicate things).  And that the authorization logic lives there and not in controller code ("guard" is probably better called "authorized").  That's the dream at least!

Here's some assumptions:
1. Authorization rules are complex and not easily expressed with simple roles
2. Current authorization rules are in Ruby and work the way we want
3. We want to avoid more complexity in authorization rules but allow it if we need it
4. Authorization rules are costly to change and will need tests and reviews

Goals
1. Small API surface area
2. Available to developers in any Ruby code without thought or effort
3. Can wrap at multiple layers

Use cases:
1. Student profile
2. Homeroom page
3. Section page
4. School overview
5. Student search bar
6. Principal dashboard
7. Restricted notes

What is authorized:
- Student
- Homeroom
- Section
- School
- Restricted Notes
(not assessments, notes, services, etc, those are all through the Student)

Performance
1. Precomputing (expire, enqueue job, lock out meanwhile)
2. For multiple calls, cache (model, educator, authorized?) per-request
3. Could introspect on Relations and optimize

**Comment by kevinrobinson** [comment]
Some more background.

### Assumptions:
1. Authorization rules are complex and not easily expressed with simple roles
2. Current authorization rules are in Ruby and work the way we want
3. We want to avoid more complexity in authorization rules but allow it if we need it
4. Authorization rules are costly to change and will need tests and reviews

### Goals
1. Small API surface area
2. Available to developers in any Ruby code without thought or effort
3. Can wrap at multiple layers
4. Striving towards not having to think about authorization when building features

### Use cases:
```
# 1) Student profile
student = authorized { Student.find(params[:id]) }`

# 2) Homeroom page
homeroom = authorized { Homeroom.(params[:id]) }
students = authorized { homeroom.students } 

# 3) Section page
section = authorized { Section.find(params[:id]) }
students = authorized { section.students }

# 4) School overview
students = authorized { School.find(params[:id]).students }

# 5) Student search bar
students = authorized { Student.all } # precomputed

# 6) Principal dashboard
students = authorized { School.find(params[:id]).students }

# 7) Restricted notes
student = authorized { Student.find(params[:id]) }
# can we add default_scope to EventNote?
```

### What gets authorized:
- Student
- Homeroom
- Section
- School
- Restricted Notes
(not assessments, notes, services, etc, those are all through the Student)

### What about performance?
This doesn't seem like a concern to me at this point, since it seems like we have a few good options for tackling it.  Those are:

1. Precomputing the value
2. For multiple `authorized` calls within the same request, we could cache authorization information in-memory per-request
3. For more intense optimizations we could introspect on lazy `ActiveRecord::Relation` objects returned by the block and optimize by doing something like fetching only what we need to authorize, filtering, then fetching the full query.

Separately, precomputing or caching anything related to authorization is usually not a great idea, but here I think the "caches all expire daily" is probably a reasonable trade-off.  If we wanted to make authorization changes in the admin panel take effect more quickly we could tackle that separately (expire the cached value, enqueue a delayed job to recompute, user is locked out until recompute is done).  All this is to say why I'm trying to avoid mixing in performance concerns here, since I think we can tackle those separately.  But I'm not sure, so more eyes and questions are great! :)

### What else?
I don't know!  Any other thoughts or feedback or ideas?

**Comment by alexsoble** [comment]
Hey @kevinrobinson! Some feedback here! In the gist it looked like we had an object-oriented approach to authorization, with classes like:

```
authorized = Authorized.new(current_educator)
authorized.students.find(student.id) # student profile page
```

The code samples you show above look more functional, with code like 

```
# 5) Student search bar
students = authorized { Student.all } # precomputed
```

The object-oriented approach was immediately obvious to me and I knew how to read it. The more functional approach is less clear and I'm not immediately sure what the `authorized` block is doing here.

**Comment by kevinrobinson** [comment]
@alexsoble There's two changes.  One is from the methods-based API to the block-based API.  That one's meaningful and we should talk about more!  The overall idea is that the block should return a model, array of models, or `ActiveRecord::Relation` and that the `authorized` method filters it out so callers only get back what an educator is authorized to see.

The change from `Authorized.new(current_educator).authorized` to `authorized` is just sugar added in the `ApplicationController` to make it less verbose, but it's not a substantive change.

**Comment by alexsoble** [comment]
@kevinrobinson I see, thanks for the explanation! Both the block-based API and the ApplicationController sugar are changes that I'm a little less familiar with but of course I want to learn about both and discuss the benefits/tradeoffs, especially for the block-based API.

**Comment by kevinrobinson** [comment]
@alexsoble Sure!  What would be the best next step?

The blocks-based API is half-baked sketched out in the comments above ([here](https://github.com/studentinsights/studentinsights/issues/1080#issuecomment-341211003) and [here](https://github.com/studentinsights/studentinsights/issues/1080#issuecomment-341525312)) and in the branch [here](https://github.com/studentinsights/studentinsights/compare/proposal/authorization#diff-af9715add29bf15dd8a7ee106f0c1c9aR36).  I'm not sure what else to write out other than just doing it and then looking together at how that comes and revising.  But I'm happy to do any other kinds of discussion or benefits/tradeoffs you think would be helpful!

**Comment by gabivoicu** [comment]
Not a super ton to add here, this makes a lot of sense to me. It is useful to have all authorization related code in on spot, both from a code organization perspective, and from a "reduce possibility for catastrophe error".

In terms of the block-based API vs method-based, is there one approach we use everywhere else in the codebase? I am very unfamiliar with benefits/tradeoffs for both, so I will add that I am always a strong proponent for consistency.

**Comment by kevinrobinson** [comment]
Thanks @gabivoicu!

I added https://github.com/studentinsights/studentinsights/pull/1237 to make it easier to discuss more concretely over there.

**Comment by kevinrobinson** [comment]
Here's some steps forward:

- [ ] ship https://github.com/studentinsights/studentinsights/pull/1274
- [ ] update seed for development to use `TestPals` in https://github.com/studentinsights/studentinsights/pull/1315
- [ ] fix authorization bug in https://github.com/studentinsights/studentinsights/issues/1304
- [ ] audit user roles and permissions in production, update `TestPals` to match
- [ ] move Educator specs related to authorization into `Authorizer` specs
- [ ] move specs for `homerooms`, `sections`, etc. into authorizer and improve them
- [ ] update call sites in each controller to call into `Authorizer` methods or to move the authorization logic in the controller into `Authorizer`.  Update corresponding tests
...later...
- [ ] consolidate code in authorizer so logic isn't redundant (eg, `is_authorized_for_student` vs. `students_for_school_overview`)

**Comment by kevinrobinson** [comment]
These steps probably need to be changed around, based on what we've learned working with this in new feature development since.  Making it safer to write code querying for `event_notes` is probably the next best step to take (see https://github.com/studentinsights/studentinsights/pull/1318#issuecomment-381102294).  We could do that next, and if this slips it'll also come up again in the next few months when pulling feature work related to the profile page, notes, etc.

**Comment by kevinrobinson** [comment]
These steps probably need to be changed around, based on what we've learned working with this in new feature development since.  Making it safer to write code querying for `event_notes` is probably the next best step to take (see https://github.com/studentinsights/studentinsights/pull/1318#issuecomment-381102294).  We could do that next, and if this slips it'll also come up again in the next few months when pulling feature work related to the profile page, notes, etc.
